adr/000x-developing-and-publishing-clientside-behaviours.md

```diff
@@ -0,0 +1,46 @@
+# 000X. Developing and Publishing ClientSide Behaviours
+
+Date: 2021-02-19
+
+## Status
+
+Proposed
+
+## Context
+
+We want to ship client side behavioural elements with Primer View Components to increase the fidelity of our components. 
+
+Doing so requires us to align to a set of constraints for the Browser ecosystem, as well as the ecosystems that consumers of Primer View Components have:
+
+ - We need to produce a JavaScript based artefact, as this is the language browsers use for client side behaviours.
+ - The artefact should be easily consumable by existing tool chains of our consumers, which may or may not be the Rails Asset Pipeline.
+ - The setup process should be uncomplicated and familiar to consumers; by using existing Rails paradigms, or JavaScript ecosystem paradigms or tools such as npm.
+ - Client side behaviours should be well tested and offer the same level of quality the existing non-behavioural components do.
+
+## Decision
+
+### TypeScript
+
+We will write behaviours in TypeScript, which is a compile-to-JavaScript language that offers stronger type guarantees leading to a higher quality level of code. TypeScript is already being used for [`primer/components`](https://github.com/primer/components) which is also a motivating factor.
+
+Behaviours will live in TypeScript files in the `app/components` directory next to their respective `.rb` files. By convention files will be named with the same basename as the Ruby counterpart. An `app/components/primer.ts` file will also exist which must import all other `.ts` files, acting as an index file. A single `.js` file with no external dependencies, which will live in `app/assets/javascripts/primer.js`. This single file can then be loaded through the Rails Asset pipeline, allowing for consumers to call `javascript_include_tag("primer")`.
```

Comment:
Since we are publishing to NPM, does it make sense to have something Rails-specific? Rails could easily 
```js
import '@primer/view_components'
```

Reply:
This document is attempting to describe offering both rails & not rails specific options. The `javascript_include_tag` allows consumers to add the primer behaviours without touching and JS, and only adding that tag to their `application.html.erb`. This means, however, that whatever is in `app/assets/javascripts/primer.js` will be loaded which may or may not include polyfills, additional libraries, so on and so forth.

In cases where more nuance is required, they can import the JS using the npm package by doing exactly what you describe, but this means they need to add their own JS file with the import and do all the necessary work to load _that_ JS file. This will use npm idiomatic conventions allowing for selective imports or bundle splitting.

I've clarified this by updating the docs.